app/utils/sparse-fieldsets.ts

```diff
@@ -0,0 +1,70 @@
+import { camelize, underscore } from '@ember/string';
+import ModelRegistry from 'ember-data/types/registries/model';
+import { pluralize, singularize } from 'ember-inflector';
+import { Links } from 'jsonapi-typescript';
+
+import { Resource } from 'osf-api';
+
+// Automatic type safety for attributes and relationships appears prohibitively expensive.
```

Comment:
What's expensive about it?

Reply:
I implemented a version with lots of type safety, where the returned `SparseModel` had types generated based on the passed-in fieldset and the corresponding fields on the models (if you're curious, [it's here](https://github.com/CenterForOpenScience/ember-osf-web/compare/release/carbon...aaxelb:sparse-fieldsets))... but it used too much memory during compilation and crashed. It's possible it could have been implemented more efficiently, but by that point I had already put in far more time than warranted by this bug.

Reply:
huh, weird